### ISSUE TITLE
Order of sum and indexing matters for GPU tensors

### ISSUE THREAD

**Issue by siddancha** [pytorch/pytorch#47168]
## 🐛 Bug

Computing the sum (along a dimension) and then indexing into a tensor should produce the same result as indexing first and then computing the sum. This holds for CPU tensors but doesn't seem to hold for GPU tensors. Is this expected behavior?

## To Reproduce

Please run `python test.py`

```python
# test.py

import torch

for i in range(1000):

    gpu_arr = torch.randn(size=[2, 800]).cuda()

    success = (gpu_arr[1].sum() == gpu_arr.sum(dim=1)[1])

    if not success:
        print("Test failed!")
        exit()

print("Test passed!")

```

## Expected behavior

I expected to get "Test passed!" but got "Test failed!" instead. The test passes if I remove `.cuda()`.

## Environment

PyTorch version: 1.5.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 440.100
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] numpy==1.17.3
[pip] torch==1.5.1
[pip] torch-cluster==1.5.5
[pip] torch-geometric==1.6.0
[pip] torch-scatter==2.0.5
[pip] torch-sparse==0.6.6
[pip] torch-spline-conv==1.2.0
[pip] torchvision==0.4.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] numpy                     1.17.3           py37hd14ec0e_0  
[conda] numpy-base                1.17.3           py37hde5b4d6_0  
[conda] torch                     1.5.1                    pypi_0    pypi
[conda] torch-cluster             1.5.5                    pypi_0    pypi
[conda] torch-geometric           1.6.0                    pypi_0    pypi
[conda] torch-scatter             2.0.5                    pypi_0    pypi
[conda] torch-sparse              0.6.6                    pypi_0    pypi
[conda] torch-spline-conv         1.2.0                    pypi_0    pypi
[conda] torchvision               0.4.2                py37_cu101    pytorch

**Comment by SsnL** [comment]
floating point sum is not associative.

**Comment by siddancha** [comment]
> floating point sum is not associative.

I see, so the order of summation is not guaranteed to be fixed on GPUs but it is on CPUs?

**Comment by SsnL** [comment]
it's fixed for same input shape and same argument, but not the same across CPU vs CUDA, nor `A[1].sum()` vs `A.sum(dim=X)[1]`.

**Comment by siddancha** [comment]
> it's fixed for same input shape and same argument, but not the same across CPU vs CUDA, nor `A[1].sum()` vs `A.sum(dim=X)[1]`.

That really helps. Thank you!!
